### DISCUSSION TITLE
TestClient - Unable to test Unhandled Errors

### DISCUSSION THREAD

**Discussion by oji** [fastapi/fastapi#9007] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

This is the code for my `main/test_main.py` file: 
```Python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from fastapi.exceptions import HTTPException

app = FastAPI()


@app.get("/")
async def read_main():
    return do_hello()

def do_hello():
    return {"msg": "Hello World"}


### Test

client = TestClient(app)


def test_read_main():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {"msg": "Hello World"}

def test_failure(mocker):
    mocked_get = mocker.patch("main.test_main.do_hello")
    mocked_get.side_effect = Exception("Something broke")
    # If using the below exception instead, it doesn't break and the assertion gets executed.
    # mocked_get.side_effect = HTTPException(500)

    response = client.get("/")
    assert response.status_code == 500
```

And I run it with `pipenv run pytest main`

### Description

I want to test a custom error handler. To do this, I mock a function that is called by my endpoint, and I expect to be able to assert that I get a 500 response with the description of the problem formatted the way I like. 

If the exception thrown is other than an `HTTPException`, the call to `client.get(...)` throws instead of returning the response generated by the custom error handler. If I use an HTTPException, I do get a response and the assertions execute, but I need this for unhandled errors as well.

### Environment

* OS: macOS
* FastAPI Version: 0.63.0
* Python version: 3.8.6

**Comment by chbndrhnns** [comment]
When I extend your snippet with an exception handler, it works for me except for the bare `Exception` which might be due to how [Starlette handles the bare exception](https://github.com/tiangolo/fastapi/issues/2683#issuecomment-764011439).

<details>
<summary>Example</summary>


```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from starlette.requests import Request
from starlette.responses import Response

app = FastAPI()


class MyError(Exception):
    ...


@app.exception_handler(MyError)
def exception(request: Request, exc: Exception):
    print("On the application:")
    print(exc)
    print("cause: ", exc.__cause__)

    return Response(status_code=500)


@app.get("/")
async def read_main():
    return do_hello()


def do_hello():
    return {"msg": "Hello World"}


### Test

client = TestClient(app)


def test_read_main():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {"msg": "Hello World"}


def test_failure(mocker):
    mocked_get = mocker.patch("app.two.do_hello")
    # mocked_get.side_effect = MyError("Something broke")
    # If using the below exception instead, it doesn't break and the assertion gets executed.
    mocked_get.side_effect = KeyError()

    response = client.get("/")
    assert response.status_code == 500

```

</details>

Is this what you were looking for?

**Comment by oji** [comment]
That's exactly the problem I'm describing: it works for me with any exception but the bare `Exception`, which is the one I'm using in the handler for unhandled errors. 

The handler for the bare `Exception` works well in the application, it just can't be tested, as it causes `client.get()` to throw rather than return a response.

So, from what I'm reading, there is no way to test a custom handler for `Exception`?

**Comment by chbndrhnns** [comment]
Looking at the linked issue, I think you might be correct about the fact that `Exception` cannot currently be handled like this.

There is a small difference between our snippets, though.
With the exception handler, you can handle ANY exception except for the bare `Exception`. 
With your snippet, I assume it would fail for ALL exceptions `HTTPException` and its sub-classes.

**Comment by alecgerona** [comment]
Had any headway with this @oji?

**Comment by nadworny** [comment]
@chbndrhnns your snippet doesn't solve the problem when you want to test how your app behaves if, for example, an unexpected error (i.e. IOError) happens. See the linked issue.

**Reply by JCHacking** [reply]
I have just faced the problem and this has been the only valid solution I have seen in this discussion.

**Comment by gusmith** [comment, marked as answer]
One thing which worked for me was to add the argument `raise_server_exceptions` which is by default set to `True` in the test client:
```
client = TestClient(app, raise_server_exceptions=False)
```
I'm using fastapi 0.68.1 which installs startlette 0.14.2, so I do not know if this option is available for earlier version...

So the whole snippet is 
```
from fastapi import FastAPI
from fastapi.testclient import TestClient
from fastapi.exceptions import HTTPException

app = FastAPI()


@app.get("/")
async def read_main():
    return do_hello()

def do_hello():
    return {"msg": "Hello World"}


### Test

client = TestClient(app, raise_server_exceptions=False)


def test_read_main():
    response = client.get("/")
    assert response.status_code == 200
    assert response.json() == {"msg": "Hello World"}

def test_failure(mocker):
    mocked_get = mocker.patch("main.test_main.do_hello")
    mocked_get.side_effect = Exception("Something broke")
    # If using the below exception instead, it doesn't break and the assertion gets executed.
    # mocked_get.side_effect = HTTPException(500)

    response = client.get("/")
    assert response.status_code == 500
```

**Comment by conradogarciaberrotaran** [comment]
I had the same issue where the exceptions were being raised when trying to do the request on the client and i realized, the exceptions handlers weren't defined in the correct place, my main.py was something like:
```python
def create_app():
   app = FastAPI()

   ...

   return app
   
@app.exception_handler(MyException)
def handle_my_exception(....):
   pass
```
my conftest.py ìs:
```python
app = create_app()
client = TestClient(app)
```
The exceptions handlers weren't being defined on the testing app instance.
Fixed it by defining the exception handlers inside the create_app function:
 ```python
def create_app():
    app = FastAPI()

    ...
    
   @app.exception_handler(MyException)
   def handle_my_exception(....):
       pass
   return app
```
Hope it helps somebody googling this error.

**Comment by deeTEEcee** [comment]
With https://github.com/tiangolo/fastapi/discussions/9007#discussioncomment-5155469 `raise_server_exceptions=False`, I ran into an issue that exceptions occuring before i even make an API call would seem to capture the result, hide it, and then I would get an unexplained 500.
